src/scheduler/gc_work.rs

```diff
@@ -683,40 +709,73 @@ impl<VM: VMBinding> ProcessEdgesWork for SFTProcessEdges<VM> {
     }
 
     fn create_scan_work(&self, nodes: Vec<ObjectReference>, roots: bool) -> ScanObjects<Self> {
-        ScanObjects::<Self>::new(nodes, false, roots)
+        ScanObjects::<Self>::new(nodes, false, roots, self.bucket)
     }
 }
-
-struct ProcessEdgesWorkRootsWorkFactory<E: ProcessEdgesWork> {
-    mmtk: &'static MMTK<E::VM>,
+struct ProcessEdgesWorkRootsWorkFactory<
+    VM: VMBinding,
+    E: ProcessEdgesWork<VM = VM>,
+    I: ProcessEdgesWork<VM = VM>,
+> {
+    mmtk: &'static MMTK<VM>,
+    phantom: PhantomData<(E, I)>,
 }
 
-impl<E: ProcessEdgesWork> Clone for ProcessEdgesWorkRootsWorkFactory<E> {
+impl<VM: VMBinding, E: ProcessEdgesWork<VM = VM>, I: ProcessEdgesWork<VM = VM>> Clone
+    for ProcessEdgesWorkRootsWorkFactory<VM, E, I>
+{
     fn clone(&self) -> Self {
-        Self { mmtk: self.mmtk }
+        Self {
+            mmtk: self.mmtk,
+            phantom: PhantomData,
+        }
     }
 }
 
-impl<E: ProcessEdgesWork> RootsWorkFactory<EdgeOf<E>> for ProcessEdgesWorkRootsWorkFactory<E> {
+impl<VM: VMBinding, E: ProcessEdgesWork<VM = VM>, I: ProcessEdgesWork<VM = VM>>
+    RootsWorkFactory<EdgeOf<E>> for ProcessEdgesWorkRootsWorkFactory<VM, E, I>
+{
     fn create_process_edge_roots_work(&mut self, edges: Vec<EdgeOf<E>>) {
         crate::memory_manager::add_work_packet(
             self.mmtk,
             WorkBucketStage::Closure,
-            E::new(edges, true, self.mmtk),
+            E::new(edges, true, self.mmtk, WorkBucketStage::Closure),
         );
     }
 
     fn create_process_node_roots_work(&mut self, nodes: Vec<ObjectReference>) {
-        // We want to use E::create_scan_work.
-        let process_edges_work = E::new(vec![], true, self.mmtk);
-        let work = process_edges_work.create_scan_work(nodes, true);
-        crate::memory_manager::add_work_packet(self.mmtk, WorkBucketStage::Closure, work);
+        crate::memory_manager::add_work_packet(
+            self.mmtk,
+            WorkBucketStage::Closure,
+            ProcessRootNode::<VM, I, E>::new(nodes, WorkBucketStage::Closure),
+        );
+    }
+
+    fn create_immovable_process_edge_roots_work(&mut self, edges: Vec<EdgeOf<E>>) {
+        crate::memory_manager::add_work_packet(
+            self.mmtk,
+            WorkBucketStage::ImmovableClosure,
+            I::new(edges, true, self.mmtk, WorkBucketStage::ImmovableClosure),
+        );
+    }
+
+    fn create_immovable_process_node_roots_work(&mut self, nodes: Vec<ObjectReference>) {
+        crate::memory_manager::add_work_packet(
+            self.mmtk,
+            WorkBucketStage::ImmovableClosure,
+            ProcessRootNode::<VM, I, I>::new(nodes, WorkBucketStage::ImmovableClosure),
```

Comment:
This is right.  Neither the packet itself nor its derived packets may move objects.